### DISCUSSION TITLE
How to interface with StarPU

### DISCUSSION THREAD

**Discussion by weslleyspereira** [tlapack/tlapack#203] [Ideas]
The question behind this discussion could be: __How would \<T\>LAPACK support optimized code in heterogeneous multicore architectures?__

@langou suggested we take a look at StarPU (https://github.com/starpu-runtime/starpu), and try to use algorithms in \<T\>LAPACK with some sort of "StarPU matrices". The premises are:
1. Algorithms in \<T\>LAPACK don't need to be modified to run with StarPU matrices. This may result in code with poor performance though. 
2. __[edited:]__ If we specialize \<T\>LAPACK kernels for variants suited with StarPU, (for example by writing "tiled algorithms", that are proven to run well over runtime,) we may achieve performance, in some cases, comparable to a code written fully in StarPU. (By specialization I mean overloading of functions)

Well, I looked at the repository of StarPU, and here are the steps we could follow in order to make the integration with StarPU happen:
1. Create a type `StarPUMatrix` with at least the operator `A(i,j)`.
2. Modify one of the examples in StarPU so that it uses `StarPUMatrix`.
3. Create a plugin in \<T\>LAPACK for `StarPUMatrix`.
4. Modify the example in step 2 so it calls \<T\>LAPACK blocked routines.
5. Specialize the kernels needed in step 4 for `StarPUMatrix`. Here we may need to make slight changes to \<T\>LAPACK blocked routines. See below.

# About the modifications in blocked routines

The example I studied is a Cholesky factorization algorithm using BLAS, namely:
```
for (k = 0; k < tiles; k++) {
  potrf(A[k,k])
  for (m = k+1; m < tiles; m++)
    trsm(A[k,k], A[m,k])
  for (m = k+1; m < tiles; m++)
    syrk(A[m,k], A[m, m])
  for (m = k+1, m < tiles; m++)
    for (n = k+1, n < m; n++)
      gemm(A[m,k], A[n,k], A[m,n])
}
```
Here, the matrix has a tiled layout and `A[i,j]` represents a block (or submatrix). I used the notation in the StarPU repository.
This example is implemented at: https://github.com/starpu-runtime/starpu/blob/63ff1b4e2723b3d03176c7505aa7fe59c7d59c6b/examples/cholesky/cholesky_implicit.c. The single-node algorithm is implemented inside the function `static int _cholesky(starpu_data_handle_t dataA, unsigned nblocks)`. Here follows a skeleton of the algorithm:

```c
static int _cholesky(starpu_data_handle_t dataA, unsigned nblocks)
{
        // Initialization

	for (k = 0; k < nblocks; k++)
	{
                // Start iteration k
		starpu_iteration_push(k);

		// Create block Akk
		// Run chol(Akk)

		for (m = k+1; m<nblocks; m++)
		{
			// Create block Amk
			// Run trsm(Akk, Amk)
		}

		// Tile Akk is ready
		starpu_data_wont_use(sdatakk);

		for (n = k+1; n<nblocks; n++)
		{
			// Create block Ank
			// Create block Ann
			// Run Ann := Ann - Ank * Ank^T

			for (m = n+1; m<nblocks; m++)
			{
				// Create block Amn
				// Create block Amn
				// Run Amn := Amn - Amk * Ank^T
			}

		        // Tile Ank is ready
			starpu_data_wont_use(sdatank);
		}
		
                // End iteration k
		starpu_iteration_pop();
	}
}
```

## Changes I foresee

__Explicitly mark a block as "ready" by calling `tlapack::ready()`__

1. The implementation of this function would be empty for matrices like `legacyMatrix`, `std::mdspan` and Eigen::Matrix, and the compiler would optimize it out. So, no problem here.
2. The implementation for `StarPUMatrix` would call `starpu_data_wont_use`.
3. Regardless, it may be a good idea to explicitly say that blocks are ready from a given point of the algorithm. It may improve readability of the blocked code.

__Mark begin and end of loops__

I am still unsure if `starpu_iteration_push` and `starpu_iteration_pop` are really needed or if they are just necessary for profiling. If it is necessary, we would need to write the most external loops in general \<T\>LAPACK blocked code as:
```c++
for (k = 0; k < nblocks; k++) {
    tlapack::start_iteration(k);
    // do the job
    tlapack::end_iteration(k);
}
```

**Comment by sthibaul** [comment]
Note: `Create block Akk` is our usual tiled algorithms is not really a creation, but just using the handle of the subdata, to give it to the task so it knows on which tile it should work.

**Comment by sthibaul** [comment]
wont_use calls indeed only provide optimization, they do not change the semantic

iteration_push/pop are indeed only useful for profiling

**Comment by weslleyspereira** [comment]
The first steps of the integration were completed in the first half of 2023. See https://github.com/tlapack/tlapack/pull/211
